File: 2972-count-the-number-of-incremovable-subarrays-ii/2972-count-the-number-of-incremovable-subarrays-ii.cpp
class Solution {
public:
    #define ll long long
    long long incremovableSubarrayCount(vector<int>& nums) {
        ll n=nums.size();
        ll l,r;
        for(l=1;l<n;++l){
            if(nums[l] <= nums[l-1]){
                break;
            }
        }
        
        for(r=n-1;r>=1;r--){
            if(nums[r] <= nums[r-1]) {
                break;
            }
        }

        l--;
        if(l>=r) return n*(n+1)/2;
        ll cnt = n-r+1;
        ll i=0,j=r;
        while(i<=l and j<n){
            if(nums[j] > nums[i]) cnt += n-j+1,i++;
            else j++;
        }
        return cnt + l-i+1;
    }
};

File: 2972-count-the-number-of-incremovable-subarrays-ii/2972-count-the-number-of-incremovable-subarrays-ii.py
class Solution:
    def incremovableSubarrayCount(self, nums: List[int]) -> int:
        nums=[float("-inf")]+nums+[float("inf")]
        n = len(nums)
        lc = 1
        rc = 1

        for i in range(1, n):
            if nums[i] <= nums[i - 1]:
                break
            lc += 1

        for i in range(n - 2, -1, -1):
            if nums[i] >= nums[i + 1]:
                break
            rc += 1

        ans=0
        for i in range(lc):
            j = bisect.bisect_left(nums, nums[i]+1, lo=max(i + 2, n - rc))
            if j < n and nums[j] > nums[i]:
                ans += n - j

        return ans